### PR TITLE
[RLlib] Add log-std clipping to 'MLPHead's.

### LINKED ISSUE #46442: [RLlib]: PPO agent training error: Invalid NaN values in Normal distribution parameters
### What happened + What you expected to happen

Hello,

I am encountering an error while training a PPO agent using RLlib. During training, I receive the following error message:
`File "/opt/conda/envs/prueba3.11/lib/python3.11/site-packages/ray/rllib/algorithms/ppo/ppo_torch_policy.py", line 85, in loss
    curr_action_dist = dist_class(logits, model)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/prueba3.11/lib/python3.11/site-packages/ray/rllib/models/torch/torch_action_dist.py", line 250, in __init__
    self.dist = torch.distributions.normal.Normal(mean, torch.exp(log_std))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/prueba3.11/lib/python3.11/site-packages/torch/distributions/normal.py", line 56, in __init__
    super().__init__(batch_shape, validate_args=validate_args)
  File "/opt/conda/envs/prueba3.11/lib/python3.11/site-packages/torch/distributions/distribution.py", line 68, in __init__
    raise ValueError(
ValueError: Expected parameter loc (Tensor of shape (128, 2)) of distribution Normal(loc: torch.Size([128, 2]), scale: torch.Size([128, 2])) to satisfy the constraint Real(), but found invalid values:
tensor([[nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan]], grad_fn=<SplitBackward0>)
`
I have checked all observations to ensure there are no NaN values, but the error persists. Can you please help me identify the cause of this issue and how to resolve it?

Thank you for your assistance.

### Versions / Dependencies

ray: 2.31
Python: 3.11
torch: 2.3.1

### Reproduction script

tuner = Tuner(
            trainable=PPO,
            param_space=...,
            run_config=...
        )
        tuner.fit()

### Issue Severity

Low: It annoys or frustrates me.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

Many implementation of continuous control algorithms suffer from instabilities in training where the log standard deviation takes on extreme values (most often very small values) and lead to numerical overflow in backward calculations (see [this discussion](https://discuss.ray.io/t/ppo-nan-in-actor-logits/15140)).

These instabilities can be partially controlled for by using a log standard deviation that can freely move and acts like a bias (i.e. it is not trained within the neural networks, but still optimized during training). Nevertheless, clipping standard deviations still is an often used technique to stabilize training further.

This PR proposes a clip parameter for the logg standard deviation and applies it in all `MLPHead`s and therefore in all algorithms that use continuous actions. More specifically:

- It introduces a new parameter `log_std_clip_param` in the `rllib/models/catalog.py`.
- It sets this parameter by default to `inf`, i.e. factually no clipping
- It passes this model config parameter into any continuous actions policy head (i.e. in `PPO`, `APPO`, `IMPALA`, `SAC`, `BC`, and `MARWIL`; note, `DreamerV3` already uses log std clipping)

## Related issue number

Closes #46442 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/models/configs.py [hunk 1]
```diff
@@ -181,6 +181,9 @@ class _MLPConfig(ModelConfig):
     output_layer_bias_initializer: Optional[Union[str, Callable]] = None
     output_layer_bias_initializer_config: Optional[Dict] = None
 
+    # Optional clip parameter for the log standard deviation.
+    log_std_clip_param: float = float("inf")
```
Comment (sven1977) [inline rllib/core/models/configs.py:185]: Could we set this to 20 by default?

File: rllib/algorithms/bc/bc_catalog.py [hunk 2]
```diff
@@ -95,6 +95,7 @@ def build_pi_head(self, framework: str) -> Model:
             hidden_layer_activation=self.pi_head_activation,
             output_layer_dim=required_output_dim,
             output_layer_activation="linear",
+            log_std_clip_param=self._model_config_dict["log_std_clip_param"],
```
Comment (sven1977) [inline rllib/algorithms/bc/bc_catalog.py:98]: What if the user doesn't define this in `model_config_dict`?

File: rllib/models/catalog.py [hunk 3]
```diff
@@ -190,6 +190,12 @@
     # outputs floating bias variables instead of state-dependent. This only
     # has an effect is using the default fully connected net.
     "free_log_std": False,
+    # Whether to clip the log standard deviation when using a Gaussian (or any
+    # other continuous control distribution). This can stabilize training and avoid
+    # very small or large log standard deviations leading to numerical instabilities
+    # which can turn network outputs to `nan`. The default is infinity, i.e. no
```
Comment (sven1977) [inline rllib/models/catalog.py:196]: "The default is to clamp the log(std) in between -20 to 20."

File: rllib/models/catalog.py [hunk 4]
```diff
@@ -190,6 +190,12 @@
...
+    # clipping.
+    "log_std_clip_param": float("inf"),
```
Comment (sven1977) [inline rllib/models/catalog.py:198]: Ah, ok, here it is. But still, as we'll soon get rid of the old stack model config dict, we should be defensive against users bringing their own `model_config_dict` to BC or other algos.

**Review by sven1977 — APPROVED** [review]
Looks good to me. Just 1-2 nits/change requests (default should be 20, not inf).

File: rllib/core/models/configs.py (hunk 1 above)
Comment (simonsays1980) [reply rllib/core/models/configs.py:185]: Computer says no ... Yes, we can :)

File: rllib/algorithms/bc/bc_catalog.py (hunk 2 above)
Comment (simonsays1980) [reply rllib/algorithms/bc/bc_catalog.py:98]: Then the default sets in doesn't it? (`rllib/models/catalog.py`)

**Pull request merged**